domain/application/state/storage.go

```diff
@@ -110,12 +379,98 @@ WHERE  storage_id = $storageInstance.storage_id
+	}
+
+	countQuery, err := st.Prepare(`
+SELECT count(*) AS &storageCount.count
+FROM storage_instance si
+JOIN storage_unit_owner suo ON si.uuid = suo.storage_instance_uuid
+WHERE suo.unit_uuid = $storageCount.unit_uuid
+AND si.storage_name = $storageCount.storage_name
+AND si.uuid != $storageCount.uuid
+`, storageCount{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// First to the basic life checks for the unit and storage.
+		unit, err := st.getMinimalUnit(ctx, tx, unitUUID)
+		if err != nil {
+			return err
+		}
+		if unit.LifeID != life.Alive {
+			return errors.Errorf("unit %q is not alive", unitUUID).Add(applicationerrors.UnitNotAlive)
+		}
+
+		stor, err := st.getStorageDetails(ctx, tx, storageUUID)
+		if err != nil {
+			return err
+		}
+		if stor.LifeID != life.Alive {
+			return errors.Errorf("storage %q is not alive", unitUUID).Add(applicationerrors.StorageNotAlive)
+		}
+
+		// See if the storage name is supported by the unit's current charm.
+		// We might be attaching a storage instance created for a previous charm version
+		// and no longer supported.
+		charmStorage, err := st.getUnitCharmStorageByName(ctx, tx, unitUUID, stor.StorageName)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf(
+					"charm for unit %q has no storage called %q",
+					unitUUID, stor.StorageName,
+				).Add(applicationerrors.StorageNameNotSupported)
+			}
+			return errors.Errorf("getting charm storage metadata for storage name %q unit %q: %w", stor.StorageName, unitUUID, err)
+		}
+
+		// Check allowed storage attachment counts - will this attachment exceed the max allowed.
+		// First get the number of storage instances (excluding the one we are attaching)
+		// of the same name already owned by this unit.
+		storageCount := storageCount{
+			StorageUUID: storageUUID,
+			StorageName: stor.StorageName,
+			UnitUUID:    unitUUID,
+		}
+		var count uint64
+		err = tx.Query(ctx, countQuery, storageCount).Get(&storageCount)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("querying storage count for storage %q on unit %q: %w", stor.StorageName, unitUUID, err)
+		}
+		if err == nil {
+			count = storageCount.Count
+		}
+		// Validate that the attachment count can increase by 1.
+		if err := validateCharmStorageCountChange(charmStorage, count, 1); err != nil {
+			return err
+		}
+
+		return st.attachStorage(ctx, tx, stor, unit, charmStorage)
+	})
+	if err != nil {
+		return errors.Errorf("attaching storage %q to unit %q: %w", storageUUID, unitUUID, err)
+	}
+	return nil
 }
 
-func (st *State) AddStorageForUnit(ctx context.Context, storageName corestorage.Name, unitUUID coreunit.UUID, stor storage.Directive) ([]corestorage.ID, error) {
+func (st *State) AddStorageForUnit(ctx context.Context, storageName corestorage.Name, unitUUID coreunit.UUID, directive storage.Directive) ([]corestorage.ID, error) {
```

Comment:
Comment me.

Reply:
I wasn't going to add comments until the implementation was done. Like for AttachStorage above.